core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+public class Context {
+
+  /**
+   * Stack of context objects which is used to record attach & detach history on a thread.
+   */
+  private static final ThreadLocal<ArrayDeque<Context>> contextStack =
```

Comment:
Wouldn't this theoretically leave a pooled thread (that might be used for some other purpose) in a dirty state? Granted it may be theoretical, since most thread pools will likely be dedicated to message processing in this case anyway. It just seems that we should clean up after ourselves, especially since there is no clear limit on how big the stack might grow ... it's entirely up to the application.